src/hash_traits.jl

```diff
@@ -0,0 +1,461 @@
+#####
+##### Helper Functions
+#####
+
+hash_trait(x::Transformer, y) = x.result_method
+hash_trait(::Transformer{<:Any,Nothing}, y) = hash_trait(y)
+hash_trait(x) = StructType(x)
+
+function check_hash_method(x, transform, context)
+    # because of how `hash_method` uses `NotImplemented` we can leverage
+    # this to check for deprecated API usage
```

Comment:
shouldn't we just deprecate all `version=2` and `version=1` calls in `stable_hash`? Rather than doing stuff like this

Reply:
My reasoning here is that it is actually quite easy to not have these deprecation warnings show up for a type that uses the implements hashing wiht the deprecated `hash_method` setup.

For example, let's say I have implemented `hash_method` for type `Foo` in package `A`, with a compat bound of "1" for `StableHashTraits`. 

Someone using `Foo` in pacakge `B` could upgrade to 1.2 and use `stable_hash(Foo(..); version=3)`. No warning will occur.

*However* unbenknownst to `B`, `A` has some customized hashing that ensures the stability of `Foo` hashes by ignoring some session dependent data that gets stored in `Foo`. However, since these are implemented using `hash_method`, the default fallbacks for `version=3` will be used, and `Foo`s hash will not be stable for `version=3`, all of which will occur silently.
```

Reply:
ah I see, so this is for `hash_method` definers, rather than `stable_hash` users, who might be different folks